Frontend/src/Components/Dashboard/GraphContainer.js

```diff
@@ -0,0 +1,258 @@
+  // the key is the name of the graph, and the value contains the selected datasets and the visible history length
+  // format: {[name: string]: {datasets: string[], historyLength: number}}
+  const [customGraphData, setCustomGraphData] = useState({});
+  // console.log("custom graph data:", customGraphData);
+
+  /**
+   * Saves the given graph's data (namely, the datasets that it is displaying and its "history length")
+   *
+   * @param {string} title the title of the graph to be saved
+   * @param {{datasets: string[], historyLength: number}} data all the data associated with the graph that is supposed to be saved
+   * @param {string[]} data.datasets the datasets that are displayed in the graph to be saved
+   * @param {number} data.historyLength the length of history, in seconds, to display to the user
+   * @param {number} index the index of the graph that's being saved
+   */
+  const onSave = useCallback(
+    (title, isNew, data, index) => {
+      // check duplicates if new
+      if (isNew && Object.keys(customGraphData).includes(title)) {
+        console.warn("No\n", title, "is in", customGraphData);
+        return false;
+      }
+
+      // set new graph data assigned to this graph
+      const tempData = Object.assign({}, customGraphData);
+      tempData[title] = data;
+      setCustomGraphData(tempData);
+
+      // update the graph title
+      const tempTitles = graphTitles.slice();
+      tempTitles[index] = title;
+      setGraphTitles(tempTitles);
+
+      // go ahead with saving
+      return true;
+    },
+    [setCustomGraphData, setGraphTitles, customGraphData, graphTitles]
+  );
+
+  //----------------- Calculate all data that can be shown --------------------
+  /**
+   * A "pre-packed" representation of the data that can be easily inserted into
+   * the graph component for rendering.
+   *
+   * The object is of the format:
+   * ```
+   * [
+   *   {
+   *     key: string,
+   *     label: string,
+   *     data: [{x: DateTime, y: any}],
+   *     borderColor: string,
+   *     backgroundColor: string
+   *   }
+   * ]
+   */
+  const packedData = useMemo(() => {
+    // console.log("memo invoked");
+    console.time("unpack graph data");
+    const temp = allDatasets.map((value) => {
+      const output = {
+        key: value.key,
+        label: value.name,
+        data: queue[value.key],
+        borderColor: value.color,
+        backgroundColor: value.color + "b3",
+      };
+      // console.log("packing:", output);
+      return output;
+    });
+    console.timeEnd("unpack graph data");
+    return temp;
+  }, [allDatasets, queue]);
+
+  // console.log("graph container propaganda");
+
+  return (
+    <VStack align="stretch" spacing={0} {...props}>
+      {graphTitles.map((graphTitle, index) => {
+        const key = graphTitle?.length
+          ? graphTitle === "Custom"
+            ? "Custom" + index
+            : graphTitles[index]
+          : index;
```

Comment:
```suggestion
            ? "Custom" + index
            : "graph_" + graphTitles[index]
          : index;
```
This allows for duplicate keys. For example, if the graph name was 1 and graphTitles[1] was empty, then there would be two keys of 1: The key from the actual graph title and the key from the index with no graph. To avoid this, you could add a fixed string to the graph title.